bin/zowe-setup-certificates.sh

```diff
@@ -0,0 +1,180 @@
+#!/bin/sh
+
+# Variables to be supplied:
+# - HOSTNAME - The hostname of the system running API Mediation
+# - IPADDRESS - The IP Address of the system running API Mediation
+# - VERIFY_CERTIFICATES - true/false - Should APIML verify certificates of services (???defaults to ???)
+# - ZOSMF_KEYRING - Name of the z/OSMF keyring
+# - ZOSMF_USER - z/OSMF server user ID
+# - EXTERNAL_CERTIFICATE - optional - Path to a PKCS12 keystore with a server certificate for APIML
+# - EXTERNAL_CERTIFICATE_ALIAS - optional - Alias of the certificate in the keystore
+# - EXTERNAL_CERTIFICATE_AUTHORITIES - optional - Public certificates of trusted CAs
+
+# - KEYSTORE_DIRECTORY - Location for generated certificates (defaults to /global/zowe/keystore???)
+# - KEYSTORE_PASSWORD - a password that is used to secure EXTERNAL_CERTIFICATE keystore and 
+#                       that will be also used to secure newly generated keystores for API Mediation.
+# - ZOWE_USER_ID - zowe user id to set up ownership of the generated certificates
+
+
+# Set up logging
+LOG_DIR=${HOME}/zowe_certificate_setup_log
+# Make the log directory if needed - first time through - subsequent installs create new .log files
+if [[ ! -d $LOG_DIR ]]; then
+    mkdir -p $LOG_DIR
+    chmod a+rwx $LOG_DIR 
+fi
+
+export LOG_FILE="certificate_config_`date +%Y-%m-%d-%H-%M-%S`.log"
+LOG_FILE=$LOG_DIR/$LOG_FILE
+touch $LOG_FILE
+chmod a+rw $LOG_FILE
+
+echo "<zowe-setup-certificates.sh>" >> $LOG_FILE
+
+# process input parameters. 
+while getopts "p:" opt; do
+  case $opt in
+    p) CERTIFICATES_CONFIG_FILE=$OPTARG;;
+    \?)
+      echo "Invalid option: -$OPTARG" >&2
+      exit 1
+      ;;
+  esac
+done
+shift "$(($OPTIND-1))"
+
+export ZOWE_ROOT_DIR=$(cd $(dirname $0)/../;pwd)
+
+# Load default values
+DEFAULT_CERTIFICATES_CONFIG_FILE=${ZOWE_ROOT_DIR}/bin/zowe-setup-certificates.env
+echo "Loading default variables from ${DEFAULT_CERTIFICATES_CONFIG_FILE} file."
+. ${DEFAULT_CERTIFICATES_CONFIG_FILE}
+
+if [[ -z ${CERTIFICATES_CONFIG_FILE} ]]
+then
+  echo "-p parameter not set. Using default ${DEFAULT_CERTIFICATES_CONFIG_FILE} file instead."
+else
+  if [[ -f ${CERTIFICATES_CONFIG_FILE} ]]
+  then
+    echo "Loading ${CERTIFICATES_CONFIG_FILE} file and overriding default variables."
+    # Load custom values
+    . ${CERTIFICATES_CONFIG_FILE}  
+  else
+    echo "${CERTIFICATES_CONFIG_FILE} file does not exist."
+    exit 1
+  fi
+fi
+
+. ${ZOWE_ROOT_DIR}/bin/zowe-init.sh 
+. ${ZOWE_ROOT_DIR}/scripts/utils/configure-java.sh
+
+# If hostname was not provided then use a default value.
+if [[ -z ${HOSTNAME} ]]
+then
+  HOSTNAME=${ZOWE_EXPLORER_HOST}
+fi
+
+# If ipaddress was not provided then use a default value.
+if [[ -z ${IPADDRESS} ]]
+then
+  IPADDRESS=${ZOWE_IP_ADDRESS}
+fi
```

Comment:
I wonder if the logic should be slightly different - zowe-init.sh will try and work out the host and ipaddress unless it already has a value, so maybe if the user can specified it should be set before init is called and if both specified not bothering to call it at all?
ie.
```
ZOWE_EXPLORER_HOST=${HOSTNAME}
ZOWE_IP_ADDRESS=${IPADDRESS}

if [[ -z ${ZOWE_EXPLORER_HOST} ]] || [[ -z ${ZOWE_IP_ADDRESS} ]]
then
	. ${ZOWE_ROOT_DIR}/bin/zowe-init.sh 
fi
. ${ZOWE_ROOT_DIR}/scripts/utils/configure-java.sh
```
Does that make sense?

Reply:
I think I need to call the zowe-init.sh script anyway because the zowe-init script also configures the ZOWE_JAVA_HOME variable. The java is required to be set for the apiml_cm.sh script. 
I could do this
```
. ${ZOWE_ROOT_DIR}/bin/zowe-init.sh 
. ${ZOWE_ROOT_DIR}/scripts/utils/configure-java.sh

if [[ ! -z ${HOSTNAME} ]]
then
  ZOWE_EXPLORER_HOST=${HOSTNAME}
fi

if [[ ! -z ${IPADDRESS} ]]
then
  ZOWE_IP_ADDRESS=${IPADDRESS}
fi
```
So that we I can write several lines below
```
SAN="SAN=dns:${ZOWE_EXPLORER_HOST},ip:${ZOWE_IP_ADDRESS},dns:localhost.localdomain,dns:localhost,ip:127.0.0.1"
```
What do you think?

Reply:
Ok - good point. The reason I wanted to set ZOWE_EXPLORER_HOST=${HOSTNAME} and ZOWE_IP_ADDRESS=${IPADDRESS} at the start is that if the ZOWE_EXPLORER_HOST and ZOWE_IP_ADDRESS values are not set in advance and zowe-init cannot find the host, or ip ((usually only happens on emulated z/OS) it prompts the user, which I don't want if they already specified it in your .env so maybe doing
```
ZOWE_EXPLORER_HOST=${HOSTNAME}
ZOWE_IP_ADDRESS=${IPADDRESS}
. ${ZOWE_ROOT_DIR}/bin/zowe-init.sh 
. ${ZOWE_ROOT_DIR}/scripts/utils/configure-java.sh
```
is the best compromise?

Reply:
Ok, I see. In that case, this is perfect!